core/embed/rust/src/ui/component/model_t1/button.rs

```diff
@@ -0,0 +1,227 @@
+}
+
+pub struct Button<T> {
+    pos: ButtonPos,
+    content: ButtonContent<T>,
+    styles: ButtonStyleSheet,
+    state: State,
+}
+
+impl<T: AsRef<[u8]>> Button<T> {
+    pub fn new(pos: ButtonPos, content: ButtonContent<T>, styles: ButtonStyleSheet) -> Self {
+        Self {
+            pos,
+            content,
+            styles,
+            state: State::Initial,
+        }
+    }
+
+    pub fn with_text(pos: ButtonPos, text: T, styles: ButtonStyleSheet) -> Self {
+        Self::new(pos, ButtonContent::Text(text), styles)
+    }
+
+    pub fn with_icon(pos: ButtonPos, image: &'static [u8], styles: ButtonStyleSheet) -> Self {
+        Self::new(pos, ButtonContent::Icon(image), styles)
+    }
+
+    pub fn enable(&mut self, ctx: &mut EventCtx) {
+        self.set(ctx, State::Initial)
+    }
+
+    pub fn disable(&mut self, ctx: &mut EventCtx) {
+        self.set(ctx, State::Disabled)
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        matches!(
+            self.state,
+            State::Initial | State::Pressed | State::Released
+        )
+    }
+
+    pub fn is_disabled(&self) -> bool {
+        matches!(self.state, State::Disabled)
+    }
+
+    pub fn content(&self) -> &ButtonContent<T> {
+        &self.content
+    }
+
+    fn style(&self) -> &ButtonStyle {
+        match self.state {
+            State::Initial | State::Released => self.styles.normal,
+            State::Pressed => self.styles.active,
+            State::Disabled => self.styles.disabled,
+        }
+    }
+
+    fn set(&mut self, ctx: &mut EventCtx, state: State) {
+        if self.state != state {
+            self.state = state;
+            ctx.request_paint();
+        }
+    }
+}
+
+impl<T: AsRef<[u8]>> Component for Button<T> {
+    type Msg = ButtonMsg;
+
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
+        match event {
+            Event::HumanInput(HidEvent::ButtonPressed(which)) => {
+                match self.state {
+                    State::Disabled => {
+                        // Do nothing.
+                    }
+                    _ => {
+                        if self.pos.hit(&which) {
+                            // Touch started in our area, transform to `Pressed` state.
+                            self.set(ctx, State::Pressed);
+                        }
+                    }
+                }
+            }
+            Event::HumanInput(HidEvent::ButtonReleased(which)) => {
+                match self.state {
+                    State::Pressed if self.pos.hit(&which) => {
+                        // Touch finished in our area, we got clicked.
+                        self.set(ctx, State::Initial);
+
+                        return Some(ButtonMsg::Clicked);
+                    }
+                    _ => {
+                        // Do nothing.
+                    }
+                }
+            }
+            _ => {}
+        };
+        None
+    }
+
+    fn paint(&mut self) {
```

Comment:
Personally i'd prefer to have the layout (bounding box) calculations outside this function.

Reply:
Yeah the layout calculations were trash, I rewrote them in c7b68d269d9641146d843a7d2035a1ffdf500edc, now they're computed top-down like in model T.